mozilla/bedrock
# Persist FxA/VPN attribution over multiple page navigations (Fixes #13147)

## One-line summary

Updates FxA/VPN attribution to store session attribution data to persist over multiple page navigations.

## Significant changes and points to review

Note for reviewer: attribution parameters will now persist for the duration of your tab remaining open (using [sessionStorage](https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage)). So it's best to close / re-open the tab after each test below:

## Issue / Bugzilla link

#13147

## Testing

Single page view tests:

- [x] UTM parameters are passed through to subscription links: https://www-demo1.allizom.org/en-US/products/vpn/?utm_source=test&utm_campaign=test
- [x] Coupon parameters is passed through to subscription links: https://www-demo1.allizom.org/en-US/products/vpn/?coupon=test
- [x] UTM and coupons are passed through to subscription links: https://www-demo1.allizom.org/en-US/products/vpn/?utm_source=test&utm_campaign=test&coupon=test

Multiple page view tests:

1. Open https://www-demo1.allizom.org/en-US/?utm_source=test&utm_campaign=test
2. Scroll to the "Featured Mozilla Products" section and click "Get Mozilla VPN"

- [x] UTM parameters from step 1 are passed through to subscription links after page navigation

WNP test (clear cookies after this test):

1. Open https://www-demo1.allizom.org/en-GB/firefox/113.0/whatsnew/
2. Click "Get Mozilla VPN"

- [x] Verify `utm_campaign=whatsnew-113-eu` is passed through to subscription links.

Opt-out test (clear cookies after this test):

1. Open https://www-demo1.allizom.org/en-US/privacy/websites/data-preferences/
2. Click "Opt out of first-party data collection"
3. Close tab
4. Open https://www-demo1.allizom.org/en-US/products/vpn/?utm_source=test&utm_campaign=test
5. Verify the `utm_source=test&utm_campaign=test` params are **not** passed through to subscription links. Only the default set of UTM params should be present (`utm_source=www.mozilla.org-vpn-product-page` and `utm_campaign=vpn-product-page`).

Opt-out test with coupon (clear cookies after this test):

1. Open https://www-demo1.allizom.org/en-US/privacy/websites/data-preferences/
2. Click "Opt out of first-party data collection"
3. Close tab
4. Open https://www-demo1.allizom.org/en-US/products/vpn/?coupon=test
5. Verify coupon is still passed, even though opt-out was clicked.
